### PR TITLE
mgr, mon: Keep upto date metadata with mgr for MONs

### PR DESCRIPTION
**Pull request by pdvian**
The mgr updates mon metadata through handle_mon_map which
gets triggered when MONs were removed/added from/to cluster or
the active mgr is restarted or mgr failsover.
We could have handled metadata update through MMgrOpen or
early MMgrReport messages but these are sent before monitor
electin completes and lead monitor updates pending metadata
in monstore. Instead of relying on fetching mon metadata using
'ceph mon metadata <id>' command, explicitly send metadata
update request with mon metadata to mgr.

Fixes: https://tracker.ceph.com/issues/55088

Signed-off-by: Prashant D <pdhange@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/mgr/DaemonServer.cc [hunk 1]
```diff
@@ -531,6 +534,49 @@ bool DaemonServer::handle_open(const ref_t<MMgrOpen>& m)
...
+    std::unique_lock locker(lock);
+
+    DaemonStatePtr daemon;
+    // Look up the DaemonState
+    if (daemon_state.exists(key)) {
+      dout(20) << "updating existing DaemonState for " << key << dendl;
+
+      daemon = daemon_state.get(key);
+      if (m->need_metadata_update == true &&
+          !m->daemon_metadata.empty()) {
```
Comment (ljflores) [inline src/mgr/DaemonServer.cc:571]: nit: the `== true` is not necessary.
```suggestion
      if (m->need_metadata_update &&
          !m->daemon_metadata.empty()) {
```

File: src/mgr/MgrClient.cc [hunk 2]
```diff
@@ -566,6 +586,31 @@ bool MgrClient::handle_command_reply(
...
+  }
+  ldout(cct,1) << service << "." << name << " metadata " << metadata << dendl;
+  service_daemon = true;
+  service_name = service;
+  daemon_name = name;
+  daemon_metadata = metadata;
+  daemon_dirty_status = true;
+
+  if (need_metadata_update == true &&
+      !daemon_metadata.empty()) {
```
Comment (ljflores) [inline src/mgr/MgrClient.cc:606]: Same here; the `== true` is not necessary.
```suggestion
  if (need_metadata_update &&
      !daemon_metadata.empty()) {
```

**Review by ljflores — APPROVED** [review]
Looks good @pdvian! I added a couple of nit comments.

File: src/mgr/DaemonServer.cc (hunk 1 above)
Comment (pdvian) [reply src/mgr/DaemonServer.cc:571]: @ljflores Agree. The reason to check m->need_metadata_update==true here is m->need_metadata_update was getting set to non-zero value during the vstart cluster testing and making this check succeed even though not intended to succeed here.

**Pull request merged**